augur/dates.py

```diff
@@ -1,65 +1,259 @@
 import argparse
 import datetime
-from textwrap import dedent
+import re
 import isodate
-import treetime.utils
+import treetime
+from textwrap import dedent
+from functools import lru_cache
+from typing import Any
+
+
+class InvalidDateFormat(ValueError):
+    pass
+
 
 SUPPORTED_DATE_HELP_TEXT = dedent("""\
     1. an Augur-style numeric date with the year as the integer part (e.g. 2020.42) or
     2. a date in ISO 8601 date format (i.e. YYYY-MM-DD) (e.g. '2020-06-04') or
-    3. a backwards-looking relative date in ISO 8601 duration format with optional P prefix (e.g. '1W', 'P1W')
+    3. an ambiguous date in ISO 8601-like format (e.g. '2020-06-XX', '2020-XX-XX') or
+    4. an incomplete date in ISO 8601-like format (e.g. '2020-06', '2020') or
+    5. a backwards-looking relative date in ISO 8601 duration format with optional P prefix (e.g. '1W', 'P1W')
 """)
 
-def numeric_date(date):
+# float, negative ok
+# note: year-only is treated as incomplete ambiguous and must be non-negative (see RE_YEAR_ONLY)
+RE_NUMERIC_DATE = re.compile(r'^-*\d+\.\d+$')
+
+# complete ISO 8601 date
+# e.g. 2018-03-25
+RE_ISO_8601_DATE = re.compile(r'^\d{4}-\d{2}-\d{2}$')
+
+# complete ambiguous ISO 8601 date
+# e.g. 2018-03-XX
+RE_AMBIGUOUS_ISO_8601_DATE = re.compile(r'^[\dX]{4}-[\dX]{2}-[\dX]{2}$')
+
+# incomplete ambiguous ISO 8601 date (missing day)
+# e.g. 2018-03
+RE_AMBIGUOUS_ISO_8601_DATE_YEAR_MONTH = re.compile(r'^[\dX]{4}-[\dX]{2}$')
```

Comment:
I guess the patterns are matching to the end of line `$` which will break on input text like "2020-10 (Day unknown)". But I assume we'll pre-process date strings to drop any parentheticals or extra spaces.

Reply:
Yeah, that seems beyond the scope of what's here so far, which only defines valid formats and converts to numeric.

As part of #860, we could have some logic for pre-processing date strings to these valid formats (e.g. `2020-10 (Day unknown)` -> `2020-10`).

Reply:
Agree on beyond the scope. Makes sense to put the pre-processing in the "clean/transform/sanitize data" #860 steps